s3parq/publish_redshift.py

```diff
@@ -0,0 +1,78 @@
+
+    reserved = [x.strip(',') for x in reserved]
+    return (name.upper() in reserved)
+
+def _validate_name(name: str):
+    if _check_reserved_keyword(name):
+        return tuple([False, f'name: {name} cannot be a reserved SQL keyword'])
+    elif not bool(re.match(r"^[a-zA-Z0-9_]", name)):
+        return tuple([False, f'name: {name} can only start with an alphanumeric or an underscore'])
+    elif bool(re.search("([ '\"])", name)):
+        return tuple([False, f'name: {name} cannot contain spaces or quotations'])
+    elif len(name) < 1 or len(name) > 127:
+        return tuple([False, f'name: {name} must be between 1 and 127 characters'])
+    else:
+        return tuple([True, None])
+
+def _redshift_name_validator(*args):
+    for arg in args:
+        response = _validate_name(arg)
+        if not response[0]:
+            raise ValueError(response[1])
+
+def _datatype_mapper(columns: dict) -> dict:
+    """ Takes a dict of column names and pandas dtypes and returns a redshift create table statement of column names/dtypes."""
+    logger.debug("Determining write metadata for publish...")
+    dtypes = {}
+    sql_statement = ""
+    for col, dtype in columns.items():
+        if dtype == 'object':
+            dtypes[col] = 'VARCHAR'
+        elif dtype.startswith('int'):
+            dtypes[col] = 'INTEGER'
+        elif dtype.startswith('float'):
+            dtypes[col] = 'REAL'
+        elif dtype.startswith('date'):
+            dtypes[col] = 'TIMESTAMP'
+        elif dtype.startswith('category'):
+            dtypes[col] = 'VARCHAR'
+        elif dtype == 'bool':
+            dtypes[col] = 'BOOLEAN'
+        else:
+            raise ValueError(f"Error: {dtype} is not a datatype which can be mapped to Redshift.")
+        sql_statement += f'{col} {dtypes[col]}, '
+    return "(" + sql_statement[:-2] + ")"
+
+def create_schema(schema_name: str, db_name: str, iam_role: str, session_helper: SessionHelper):
+    _redshift_name_validator(schema_name, db_name)
+    with session_helper.db_session_scope() as scope:
+        new_schema_query = f"CREATE EXTERNAL SCHEMA IF NOT EXISTS {schema_name} \
+                FROM DATA CATALOG \
+                database '{db_name}' \
+                iam_role '{iam_role}' \
+                CREATE EXTERNAL DATABASE IF NOT EXISTS;"
+
+        logger.info(f'Running query to create schema: {new_schema_query}')
+        scope.execute(new_schema_query)
+
+def create_table(table_name: str, schema_name: str, columns: dict, partitions: dict, path: str, session_helper: SessionHelper):
+    _redshift_name_validator(table_name)
+    redshift_columns = _datatype_mapper(columns)
```

Comment:
I love that we can use the same function for generating the `redshift_columns` and `redshift_partitions`